### PR TITLE
minmax: fix copy index data in a wrong way

### LINKED ISSUE #8301: Copy rough set index Data in a wrong way
## Bug Report

Please answer these questions before submitting your issue. Thanks!

https://github.com/pingcap/tiflash/blob/ec00c9ae143b5b8ef7582972e5fee5eb5a4fbf39/dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp#L307-L311

When using String, we should use reinterpret_cast<const char *>(&chars[prev_offset]) instead of chars[prev_offset], so that it will call constructor `constexpr basic_string( const CharT* s, const Allocator& alloc = Allocator())` rather than `constexpr basic_string( size_type count, CharT ch, const Allocator& alloc = Allocator() );`

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: close #8301

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp [hunk 1]
```diff
@@ -305,10 +305,10 @@ RSResults MinMaxIndex::checkNullableEqual(
             size_t pos = i * 2;
             size_t prev_offset = pos == 0 ? 0 : offsets[pos - 1];
             // todo use StringRef instead of String
-            auto min = String(chars[prev_offset], offsets[pos] - prev_offset - 1);
+            auto min = String(reinterpret_cast<const char *>(&chars[prev_offset]), offsets[pos] - prev_offset - 1);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp:308]: Add some description about why we **must** add this cast

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp:311]: Added.

**Review by CalvinNeo — APPROVED** [review]
lgtm

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**